design/design-doc-composition.md

```diff
@@ -0,0 +1,1079 @@
+Crossplane will automatically create a `kind: CustomResourceDefinition`, that
+allows application operators to author the below custom resource:
+
+```yaml
+apiVersion: database.example.org/v1alpha1
+kind: MySQLInstance
+metadata:
+  name: sql
+spec:
+  # The schema for the following three fields is defined by the above
+  # InfrastructureDefinition.
+  engineVersion: "5.7"
+  storageGB: 10
+  region: us-west
+  # The infrastructure object schema is automatically injected into the
+  # CustomResourceDefinition that is created by the InfrastructureDefinition
+  # that defines the MySQLInstance resource.
+  infrastructure:
+    # Multiple compositions may potentially satisfy a particular kind of
+    # infrastructure. Exactly one composition may be annotated as the default
+    # for a particular kind of infrastructure. Each infrastructure instance may
+    # influence which composition is used via label selectors. This could be
+    # used, for example, to determine whether a GCP CloudSQLInstance or an Azure
+    # SQLServer based composition satisfied this MySQLInstance.
+    compositionSelector:
+     matchLabels:
+       connectivity: private
+    # The MySQLInstance author may explicitly select which composition should be
+    # used by setting the compositionRef. In the majority of cases the author
+    # will ignore this field and it will be set by a controller, similar to the
+    # contemporary classRef field.
+    compositionRef:
+    - name: private-mysql-server
+    # Each infrastructure resource maintains an array of the resources it
+    # composes. Composed resources are always cluster scoped, and always either
+    # primitive or composite infrastructure resources. Composed resources model
+    # their relationship with the infrastructure resource via their controller
+    # reference. The infrastructure resource must maintain this array because
+    # there is currently no user friendly, performant way to discover which
+    # resources (of arbitrary kinds) are controlled by a particular resource per
+    # https://github.com/kubernetes/kubernetes/issues/54498
+    composedRefs:
+    - apiVersion: azure.crossplane.io/v1alpha3
+      kind: ResourceGroup
+      name: sql-34jd2
+    - apiVersion: database.azure.crossplane.io/v1beta1
+      kind: MySQLServer
+      name: sql-3i3d1
+    - apiVersion: database.azure.crossplane.io/v1alpha3
+      kind: MySQLServerVirtualNetworkRule
+      name: sql-2mdus
+    # The MySQLInstance author must specify where the MySQLInstance will write
+    # its connection details as a Kubernetes secret. The keys of the secret are
+    # specified by the InfrastructureDefinition.
+    writeConnectionSecretToRef:
+      namespace: crossplane-system
+      name: sql
+    # This cluster scoped MySQLInstance _may_ bind to exactly one namespaced
+    # MySQLInstanceBinding. See InfrastructurePublication below for details.
+    binding:
```

Comment:
Since the relation is 1-to-1, why not do:
```yaml
bindingRef:
  apiVersion: database.example.org/v1alpha1
  kind: MySQLInstanceBinding
  namespace: default
  name: sql
```

Reply:
I think the only reason it was `spec.binding.ref` instead of `spec.bindingRef` was to group the `reclaimPolicy` under `spec.binding.reclaimPolicy`, since the reclaim policy relates to the lifecycle of the binding. I'm happy to go for `spec.bindingRef` and `spec.reclaimPolicy` instead though.